### ISSUE TITLE
Helper - GIS positions need to be supplied as Long/Lat

### ISSUE THREAD

**Issue by IanMayo** [debrief/pepys-import#142]
We're currently creating the Location geometry elements in the parser code, and pushing a Geometry object to the `State`.

The above allows the parser programmer to inadvertantly provide Lat/Long, whereas PostGIS expects Long/Lat.

We're also creating/managing Heading/Course/Speed with typed units, then pushing the plain double value to the `State`.

The above would still allow heading to be provided in Degrees, or Speed in km/hour.

We can _hold the hand_ of the parser maintainer if we introduce more advanced setter functions for these fields.

E.g. If `state.setLocation(lat: double, lon: double)` will hide the _swapping_ from the parser  maintainer, and `state.setLocation(latDegs, latMins, latSecs, latHem, longDegs, longMins, longSecs, longHem)` could ease the task of collating the data.  

Similarly, `state.setSpeed(speed: quantity)` will hide the internal units from the parser maintainer.  Within `setSpeed` we may even be able to verify that it's a type of unit that's capable of being converted to the internal units (m/sec). 

So, our advanced setters would take higher level objects, and ensure the internal (temporary) field was created of the correct type/units.

**Comment by BarisSari** [comment]
I think it only makes sense to have setters if we are going to check the type of the field. Because `method(a: float, b: str):` actually doesn't mean anything to Python. Type hints are only caught by IDEs or libraries [(link)](https://docs.python.org/3/library/typing.html). However, as you mentioned, we can verify its type. For example, we have a utility function _**(convert_speed)**_ that converts float speed value meter/seconds right now. We first try to convert the input to float, if it's successful, we convert it to m/s and set the speed value of `State`. We also have a `Location` class for converting degrees-minutes-seconds-hemisphere style locations. We can insert verification codes there as well.

To sum up, I believe that we already hide the internal units from the parser maintainer. We call functions or classes, send tokens(string, float, etc.) and get back the related data in the correct unit. So, I think it may be enough if we only improve them. But if they are not enough, we can think about setter functions.

**Comment by IanMayo** [comment]
> Note: I think this answer explains why setters-getters are not Pythonic way.

Yes, but it appears a setter can be _hidden_ as an attribute like this:
```
    @property
    def attribute(self): # implements the get - this name is *the* name
        return self._attribute
```

Getting the units wrong is a real challenge for our system, and the parser maintainers are the _weak link_, since they may come & go, & not be familiar with the architectural decisions we've made.

Problems with using the [wrong units](https://www.wired.com/2010/11/1110mars-climate-observer-report/).

So, if in the _parser space_, speeds, distances, angles, time periods are managed as Quantities, any value will _carry its units with it_.   We can enforce that by making sure that any time one of the dimensioned attributes is set on a measurement - it has its units with it.  

It's possible for a parser maintainer to _accidentally_ do `state.setSpeed(double)`.  But, if we have a _hidden setter_, we can throw an error if anything other than a quantity-with-units is passed in.  To generate the quantity-with-units, the parser maintainer is going to have to explicitly state (and maybe research first) the units for that measurement.

The clients do have datasets where latitude and longitude are expressed in radians, or as 0..1.  That brings an argument for `state.location = (lat: lat_in_degs, long: long_in_degs)` (or some pythonic way of doing that :-) )

@robintw  - do you have a scientific/academic perspective on this?

**Comment by robintw** [comment]
I agree that the units are very important, and with potentially inexperienced people writing the parsers we need to make sure we deal with them in the most robust way possible.

In my mind, this means that all quantities in the parsers themselves should be converted to Quantities as soon as possible. So, the flow in the parser would look a bit like this:

```
# Get the speed token from the list of tokens for this line
speed_token = tokens[3]
# Try and parse the speed as a float
speed = parseSpeed(speed_token)
# Explicitly convert the speed to a Quantity
speed = speed * units.knots
# Record the extraction
speed_token.record(parser_name, speed.magnitude, "speed", "knots")
# Set the speed on the state
state.setSpeed(speed)
```

The annoyance here is that we have two separate things that need doing: one is converting the speed into a float (and possibly doing some basic validation on it - like checking > 0 and < 360 for heading) and another is setting the speed for the State object. In the example I've separated out those steps, so that we explicitly set the units in the parser. It would be possible to do something like

```
speed = parseSpeed(speed_token, units.knots)
```

but I'm not sure which is best. The other question becomes whether we provide these functions like `parseSpeed` or `parseHeading` (a bit like `convert_absolute_angle` that we have already), or expect the user to write ones that are relevant to their parser (as there may need to be subtle difference between them for different parsers - which would get very confusing).

As for the question of attributes vs getters/setters...I'm going to do something _very_ rare for me, and actually argue that we should use setters rather than attributes (I usually hate this idea!). However, in this case I think it is cleaner, mainly because a number of times we will be setting with multiple arguments, and that's _really_ messy to do with attributes (Ian's example of setting an attribute to a dictionary is an example of this).

As an example, for setting the location we may want to accept various different sets of arguments:

```
setLocation(latitude_decimal_degrees, longitude_decimal_degrees)
setLocation(lat_deg, lat_min, lat_sec, lon_deg, lon_min, lon_sec)
```

Those would be very hard to do as an attribute, but much cleaner as a set of setter functions. Unfortunately Python doesn't have method overloading or multiple dispatch, so we have a few options:

1. Just have multiple functions where needed: `setLocation_decimal`, `setLocation_dms`. (Simple, but has more for the user to remember)

2. Have one function and do some complicated things with keyword arguments which means if it only gets two arguments then it treats it as decimal degrees, if it has six then it treats it as dms (possibly cleaner, but more complicated - and the list of arguments and docstring would look _weird_ and confusing)

3. Use a library like [multipledispatch](https://pypi.org/project/multipledispatch/) (see [good example here](https://stackoverflow.com/questions/6434482/python-function-overloading) to allow multiple functions to be called depending on the arguments (like proper function overloading/multiple dispatch). (cleanest, but also most complicated)

Let me know what you think

**Comment by robintw** [comment]
I think these changes are probably a sensible thing for me to focus on soon - after the highlighter code is merged. What do you think about my suggestions and questions above @IanMayo @BarisSari?

**Comment by IanMayo** [comment]
Yes, @robintw - good idea.  Yes, it's a good, tidy piece of work.   

Note: there is a corollary aspect to ensuring the import code correctly handles units.  We retrieve data from the database in the Export code.  There won't be a lot of export code - it may not go beyond `Export to REPLAY` format.  But, as we're providing setters that assume a typed quantity (when relevant), we should provided getters that return that measurement attribute in a typed quantity.  In the database, we're storing data in SI units - so the getter may return the attribute in different units to what it was originally measured as.  But, the person writing the export code will then consciously convert the attribute to the units they require for the export format.

Different parsers will have different implementations of `parseSpeed` - so we'll have to leave that to the parser writers.  But, I think we'll be encouraging them to good data-units practice by forcing them to create a typed quantity in order to call `state.setSpeed(speed_with_units)`.

Lastly, we have the strategy for support of multiple ways of expressing locations.  I'm ok with location being a _special case_ - a composite structure.  If we create a class that represents location, it can have a series of create/init constructors that take different numbers of parameters, but store double lon/lat internally.  It will be objects of this type that get passed to `state.setLocation(location)`, or returned from `state.getLocation()`

**Comment by robintw** [comment]
Just realised that before doing this, I will need to fix https://github.com/debrief/pepys-import/issues/170 or I will end up having to duplicate this functionality for the SQLite and Postgres classes.

**Comment by IanMayo** [comment]
@robintw - we just chatted about this.

I now recall earlier in the day seeing the `dimensionality` of a typed field:
https://pint.readthedocs.io/en/0.10.1/tutorial.html

I _think_ this is what we would need to verify inside `setRange(range)` -  The `range` object should be a pint quantity, and it should have a `dimensionality` of `[length]`

(yes, I chose `range` since I was clear on the dimensions for it 🤣 .  I _think_ angles are dimensionless)

**Comment by robintw** [comment]
Fixed in #206, #224 and #233
